File: SBC-2022/problema_m.py
from time import time

def maxComumDivisor(x, y):
    if(y==0):
        return x
    else:
        return maxComumDivisor(y,x%y)

n, q = (int(x) for x in input().split(" "))

jogadores = [int(x) for x in input().split(" ")]
casaZero = [-1] * len(jogadores)

st = time()

for j in range(q):
    cq, dq = (int(x) for x in input().split(" "))
    for i in range(n):
        if(jogadores[i] > 0):
            if(maxComumDivisor(i+1, cq) > 1):
                jogadores[i] -= dq
                if(jogadores[i] <= 0):
                    casaZero[i] = j+1
    
    # print(jogadores)
    # print(casaZero)

for final in casaZero:
    print(final)

    
et = time()
# get the execution time
elapsed_time = (et - st)
print('Execution time:', elapsed_time, 'seconds')




# from time import time

# def maxComumDivisor(x, y):
#     if(y==0):
#         return x
#     else:
#         return maxComumDivisor(y,x%y)

# n, q = (int(x) for x in input().split(" "))

# jogadores = [int(x) for x in input().split(" ")]
# casaZero = [-1] * len(jogadores)

# st = time()

# tam = n

# for j in range(q):
#     cq, dq = (int(x) for x in input().split(" "))
#     i = 0
#     k = 0
#     while(i < tam):
#         if(jogadores[i] > 0):
#             if(maxComumDivisor(k+1, cq) > 1):
#                 jogadores[i] -= dq
#                 if(jogadores[i] <= 0):
#                     casaZero[i] = j+1
#                     del jogadores[i]
#                     tam-=1
#                 else:
#                     i+=1
#             else:
#                 i+=1
#         else:
#             i+=1
#         k += 1

# for final in casaZero:
#     print(final)

# et = time()
# # get the execution time
# elapsed_time = (et - st)
# print('Execution time:', elapsed_time, 'seconds')

File: SBC-2022/problema_m.cpp
#include <iostream>
#include <vector>

using namespace std;

int maxComumDivisor(int x, int y)
{
    if (y == 0)
    {
        return x;
    }
    else
    {
        return maxComumDivisor(y, x % y);
    }
}

void optimize()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
}

int main()
{

    optimize();

    long long n, q;
    long long cq, dq, val;
    long long j, i, k;

    cin >> n >> q;

    vector<long long> jogadores;
    vector<long long> pos;
    vector<long long> casaZero;

    for (i = 0; i < n; i++)
    {
        cin >> val;
        jogadores.push_back(val);
        pos.push_back(i);
        casaZero.push_back(-1);
    }

    long long tam = n;

    for (j = 0; j < q; j++)
    {
        cin >> cq >> dq;
        i = 0;
        while (i < tam)
        {
            if (maxComumDivisor(pos[i] + 1, cq) > 1)
            {
                jogadores[i] -= dq;
                if (jogadores[i] <= 0)
                {
                    casaZero[pos[i]] = j + 1;
                    jogadores.erase(jogadores.begin() + i);
                    pos.erase(pos.begin() + i);
                    tam--;
                }
                else
                {
                    i++;
                }
            }
            else
            {
                i++;
            }
            k++;
        }
        // for (i = 0; i < n; i++)
        // {
        //     cout << casaZero[i] << " | ";
        // }
    }

    for (i = 0; i < n; i++)
    {
        cout << casaZero[i] << "\n";
    }

    return 0;
}

// for (i = 0; i < n; i++)
// {
//     if (jogadores[i] > 0)
//     {
//         if (maxComumDivisor(i + 1, cq) > 1)
//         {
//             jogadores[i] -= dq;
//             if (jogadores[i] <= 0)
//             {
//                 casaZero[i] = j + 1;
//             }
//         }
//     }
// }